app/src/main/java/org/wikipedia/login/LoginActivity.kt

```diff
@@ -0,0 +1,265 @@
+import org.wikipedia.util.DeviceUtil.hideSoftKeyboard
+import org.wikipedia.util.FeedbackUtil.showError
+import org.wikipedia.util.FeedbackUtil.showMessage
+import org.wikipedia.util.FeedbackUtil.showPrivacyPolicy
+import org.wikipedia.util.UriUtil.visitInExternalBrowser
+import org.wikipedia.util.log.L.w
+import org.wikipedia.views.NonEmptyValidator
+
+class LoginActivity : BaseActivity() {
+    private var firstStepToken: String? = null
+    private var funnel: LoginFunnel? = null
+    private var loginSource: String? = null
+    private val loginClient = LoginClient()
+    private val loginCallback: LoginCallback = LoginCallback()
+    private var shouldLogLogin = true
+    private lateinit var binding: ActivityLoginBinding
+    public override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        binding = ActivityLoginBinding.inflate(layoutInflater)
+        binding.viewLoginError.backClickListener = View.OnClickListener { onBackPressed() }
+        binding.viewLoginError.retryClickListener = View.OnClickListener { binding.viewLoginError.visibility = View.GONE }
+
+        // Don't allow user to attempt login until they've put in a username and password
+        NonEmptyValidator(binding.loginButton, binding.loginUsernameText, binding.loginPasswordInput)
+        binding.loginPasswordInput.editText?.setOnEditorActionListener { _, actionId, _ ->
+            if (actionId == EditorInfo.IME_ACTION_DONE) {
+                validateThenLogin()
+                return@setOnEditorActionListener true
+            }
+            false
+        }
+        funnel = LoginFunnel(WikipediaApp.getInstance())
+        loginSource = intent.getStringExtra(LOGIN_REQUEST_SOURCE)
+        if (!TextUtils.isEmpty(loginSource) && loginSource == LoginFunnel.SOURCE_SUGGESTED_EDITS) {
+            Prefs.setSuggestedEditsHighestPriorityEnabled(true)
+        }
+
+        // always go to account creation before logging in
+        if (savedInstanceState == null) {
+            startCreateAccountActivity()
+        }
+        setAllViewsClickListener()
+
+        // Assume no login by default
+        setResult(RESULT_LOGIN_FAIL)
+    }
+
+    private fun setAllViewsClickListener() {
+        binding.loginButton.setOnClickListener { validateThenLogin() }
+        binding.loginCreateAccountButton.setOnClickListener { startCreateAccountActivity() }
+        binding.inflateLoginAndAccount.privacyPolicyLink.setOnClickListener { showPrivacyPolicy(this) }
+        binding.inflateLoginAndAccount.forgotPasswordLink.setOnClickListener {
+            val title = PageTitle("Special:PasswordReset", WikipediaApp.getInstance().wikiSite)
+            visitInExternalBrowser(this, Uri.parse(title.uri))
+        }
+    }
+
+    private fun getText(input: TextInputLayout): String {
+        return StringUtils.defaultString(if (input.editText != null && input.editText!!.text != null) input.editText!!.text.toString() else "")
```

Comment:
All instances of `StringUtils.defaultString()` can be replaced with kotlin's built-in `[string?].orEmpty()`